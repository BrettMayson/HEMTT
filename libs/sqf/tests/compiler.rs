#![allow(clippy::unwrap_used)]

pub use float_ord::FloatOrd as Scalar;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{compiler::serializer::Compiled, parser::database::Database};
use hemtt_workspace::LayerType;

macro_rules! compile {
    ($dir: expr, $file:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $file>]() {
                let (bin, compiled) = optimize($dir, stringify!($file));
                // bin to hex
                let hex = bin.iter().map(|b| format!("{:02x}", b)).collect::<String>();
                insta::assert_snapshot!(hex);
                insta::assert_debug_snapshot!(compiled);
            }
        }
    };
}

compile!("optimizer", consume_array);
compile!("optimizer", static_math);
compile!("optimizer", scalar);
compile!("optimizer", string_case);
compile!("simple", dev);
compile!("simple", eventhandler);
compile!("simple", foreach);
compile!("simple", format_font);
compile!("simple", get_visibility);
compile!("simple", hash_select);
compile!("simple", hello);
compile!("simple", include);
compile!("simple", oneline);
compile!("simple", semicolons);

const ROOT: &str = "tests/";

fn optimize(folder: &str, file: &str) -> (Vec<u8>, Compiled) {
    let folder = std::path::PathBuf::from(ROOT).join(folder);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join(format!("{file}.sqf")).unwrap();
    let processed = Processor::run(&source).unwrap();
    let mut writer = Vec::new();
    let statements = hemtt_sqf::parser::run(&Database::a3(false), &processed).unwrap();
    let compiled = statements.compile(&processed).unwrap();
    statements
        .optimize()
        .compile_to_writer(&processed, &mut writer)
        .unwrap();
    (writer, compiled)
}
