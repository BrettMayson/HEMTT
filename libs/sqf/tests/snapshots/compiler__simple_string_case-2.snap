---
source: libs/sqf/tests/compiler.rs
expression: compiled
---
Compiled {
    entry_point: 5,
    constants_cache_compression: true,
    constants_cache: [
        String(
            "A",
        ),
        String(
            "b",
        ),
        String(
            "C",
        ),
        String(
            "d",
        ),
        String(
            "#line 1 \"\\string_case.sqf\"\ntoLower \"A\" + toUpper \"b\" + toUpperAnsi \"C\" + toLowerAnsi \"d\";\n",
        ),
        Code(
            Instructions {
                contents: [
                    EndStatement,
                    Push(
                        0,
                    ),
                    CallUnary(
                        0,
                        SourceInfo {
                            offset: 27,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    Push(
                        1,
                    ),
                    CallUnary(
                        1,
                        SourceInfo {
                            offset: 41,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    CallBinary(
                        2,
                        SourceInfo {
                            offset: 39,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    Push(
                        2,
                    ),
                    CallUnary(
                        3,
                        SourceInfo {
                            offset: 55,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    CallBinary(
                        2,
                        SourceInfo {
                            offset: 53,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    Push(
                        3,
                    ),
                    CallUnary(
                        4,
                        SourceInfo {
                            offset: 73,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                    CallBinary(
                        2,
                        SourceInfo {
                            offset: 71,
                            file_index: 0,
                            file_line: 1,
                        },
                    ),
                ],
                source_pointer: Constant(
                    4,
                ),
            },
        ),
    ],
    names_cache: [
        "tolower",
        "toupper",
        "+",
        "toupperansi",
        "toloweransi",
    ],
    file_names: [
        "/string_case.sqf",
    ],
}
